File: surimi/Parser.cs
namespace Surimi;

using System.Collections.Generic;

public class Parser {
    public Parser(IEnumerable<Token> tokens, ErrorReporter onError)
    {
        _tokens = new PeekableStream(tokens);
        _onError = onError;
    }

    public static List<Stmt>? Parse(IEnumerable<Token> tokens,
      ErrorReporter onError)
    {
        var p = new Parser(tokens, onError);
        var prog = new List<Stmt>();
        bool failed = false;
        while (!p.AtEOF) {
            try {
                prog.Add(p.Declaration());
            } catch (ParseError) {
                failed = true;
                p.Recover();
            }
        }
        return failed ? null : prog;
    }

    private Stmt Declaration()
    {
        Token? tok;
        if ((tok = Match(TokenType.Var)) != null)
            return VarDeclRest(tok.Value);
        if ((tok = Match(TokenType.Fun)) != null)
            return FunDefOrMethodRest(tok.Value);
        if ((tok = Match(TokenType.Class)) != null)
            return ClassDefRest(tok.Value);
        return Statement();
    }

    private VarDecl VarDeclRest(Token var_)
    {
        var name = Require("expected identifier", TokenType.Ident);
        Expr? initializer = null;
        if (Match(TokenType.Eq) != null) {
            initializer = Expression();
        }
        Require("expected ';'", TokenType.Semicolon);
        return new VarDecl(
          new Var(name.Lexeme, name.Location), initializer, var_.Location);
    }

    private FunDef FunDefOrMethodRest(Token tok)
    {
        string kind;
        Var name;
        switch (tok.Type) {
            case TokenType.Fun:
                kind = "function";
                var ident = Require("expected function name", TokenType.Ident);
                name = new Var(ident.Lexeme, ident.Location);
                break;
            case TokenType.Ident:
                kind = "method";
                name = new Var(tok.Lexeme, tok.Location);
                break;
            default:
                throw new InvalidOperationException(
                  "internal error: bad token type for fun/method rest");
        }

        Require($"expected '(' after {kind} name", TokenType.LParen);

        List<Var> parameters = new List<Var>();
        if (!Check(TokenType.RParen)) {
            do {
                var ident = Require("expected parameter name", TokenType.Ident);
                parameters.Add(new Var(ident.Lexeme, ident.Location));
            } while (Match(TokenType.Comma) != null);
        }

        var rparen = Require("expected ')'", TokenType.RParen);
        if (parameters.Count > 255)
            _onError.Error(rparen.Location, $" at {rparen.Lexeme}",
              "more than 255 arguments");

        var lbrace = Require($"expected '{{' before {kind} body",
          TokenType.LBrace);
        var body = BlockRest(lbrace);

        return new FunDef(name, parameters, body.Statements, tok.Location);
    }

    private ClassDef ClassDefRest(Token tok)
    {
        Token name = Require("expected class name", TokenType.Ident);
        Var? super = null;
        if (Match(TokenType.Lt) != null) {
            Token superName = Require("superclass name", TokenType.Ident);
            super = new Var(superName.Lexeme, superName.Location);
        }
        Require("expected '{'", TokenType.LBrace);
        List<FunDef> methods = new List<FunDef>();
        while (!Check(TokenType.RBrace) && !AtEOF) {
            var methodName = Require("expected method name", TokenType.Ident);
            methods.Add(FunDefOrMethodRest(methodName));
        }
        Require("expected '}'", TokenType.RBrace);
        return new ClassDef(new Var(name.Lexeme, name.Location),
          super, methods, tok.Location);
    }

    private Stmt Statement()
    {
        Token? tok;
        if ((tok = Match(TokenType.If)) != null)
            return IfElseRest(tok.Value);
        if ((tok = Match(TokenType.While)) != null)
            return WhileRest(tok.Value);
        if ((tok = Match(TokenType.For)) != null)
            return ForRest(tok.Value);
        if ((tok = Match(TokenType.Print)) != null)
            return PrintRest(tok.Value);
        if ((tok = Match(TokenType.Return)) != null)
            return ReturnRest(tok.Value);
        if ((tok = Match(TokenType.LBrace)) != null)
            return BlockRest(tok.Value);

        return ExpressionStatement();
    }

    private IfElse IfElseRest(Token if_)
    {
        Require("expected '('", TokenType.LParen);
        var cond = Expression();
        Require("expected ')'", TokenType.RParen);
        Stmt s_if = Statement();
        Stmt? s_else = null;
        if (Match(TokenType.Else) != null)
            s_else = Statement();
        return new IfElse(cond, s_if, s_else, if_.Location);
    }

    private While WhileRest(Token while_)
    {
        Require("expected '('", TokenType.LParen);
        var cond = Expression();
        Require("expected ')'", TokenType.RParen);
        var body = Statement();
        return new While(cond, body, while_.Location);
    }

    private Stmt ForRest(Token for_)
    {
        Require("expected '('", TokenType.LParen);

        Token? init_tok = null;
        Stmt? initializer;
        if (Match(TokenType.Semicolon) != null)
            initializer = null;
        else if ((init_tok = Match(TokenType.Var)) != null)
            initializer = VarDeclRest(init_tok.Value);
        else
            initializer = ExpressionStatement();

        Token? cond_tok = null;
        Expr? condition;
        if ((cond_tok = Match(TokenType.Semicolon)) != null) {
            condition = null;
        } else {
            condition = Expression();
            Require("expected ';'", TokenType.Semicolon);
        }

        Expr? increment;
        if (Match(TokenType.RParen) != null) {
            increment = null;
        } else {
            increment = Expression();
            Require("expected ')'", TokenType.RParen);
        }

        var body = Statement();

        if (increment != null)
            body = new Block(new List<Stmt> {
                body,
                new ExprStmt(increment, increment.Location),
            }, body.Location);

        if (condition == null)
            // point to the ;
            condition = new Literal(true, cond_tok!.Value.Location);

        body = new While(condition, body, for_.Location);

        if (initializer != null)
            return new Block(new List<Stmt> {
                initializer,
                body,
            }, initializer.Location);

        return body;
    }

    private Print PrintRest(Token print)
    {
        var e = Expression();
        Require("expected ';'", TokenType.Semicolon);
        return new Print(e, print.Location);
    }

    private Return ReturnRest(Token return_)
    {
        if (Match(TokenType.Semicolon) != null)
            return new Return(null, return_.Location);
        var e = Expression();
        Require("expected ';'", TokenType.Semicolon);
        return new Return(e, return_.Location);
    }

    private Block BlockRest(Token lbrace)
    {
        var stmts = new List<Stmt>();
        while (!Check(TokenType.RBrace) && !AtEOF)
            stmts.Add(Declaration());
        Require("expected '}'", TokenType.RBrace);
        return new Block(stmts, lbrace.Location);
    }

    private ExprStmt ExpressionStatement()
    {
        var e = Expression();
        Require("expected ';'", TokenType.Semicolon);
        return new ExprStmt(e, e.Location);
    }

    private Expr Expression()
    {
        return Assignment();
    }

    private Expr Assignment()
    {
        var e = LogicOr();

        Token? eq;
        if ((eq = Match(TokenType.Eq)) != null) {
            var val = Assignment();
            switch (e) {
                case Var v:
                    return new Assign(v, val, v.Location);
                case PropertyGet g:
                    return new PropertySet(g.Object, g.Name, val, g.Location);
                default:
                    _onError.Error(eq.Value.Location,
                      $" at '{eq.Value.Lexeme}'",
                      $"invalid assignment target: {e.PrettyPrint()}");
                    throw new ParseError();
            }
        }

        return e;
    }

    private delegate Expr ExprParser();

    private Expr BinOpFold(ExprParser parser, params TokenType[] types)
    {
        var lhs = parser();
        Token? op_tok;
        while ((op_tok = Match(types)) != null) {
            var rhs = parser();
            lhs = new BinOpApp(TokenBinOp(op_tok.Value), lhs, rhs,
              op_tok.Value.Location);
        }
        return lhs;
    }

    private Expr LogicOr() => BinOpFold(LogicAnd, TokenType.Or);

    private Expr LogicAnd() => BinOpFold(Equality, TokenType.And);

    private Expr Equality() => BinOpFold(Comparison,
      TokenType.NotEq, TokenType.EqEq);

    private Expr Comparison() => BinOpFold(Term,
      TokenType.Gt, TokenType.GtEq, TokenType.Lt, TokenType.LtEq);

    private Expr Term() => BinOpFold(Factor, TokenType.Plus, TokenType.Minus);

    private Expr Factor() => BinOpFold(Unary, TokenType.Star, TokenType.Slash);

    private Expr Unary()
    {
        Token? op_tok = Match(TokenType.Not, TokenType.Minus);
        if (op_tok != null) {
            var rhs = Unary();
            return new UnOpApp(TokenUnOp(op_tok.Value), rhs,
              op_tok.Value.Location);
        }
        return Call();
    }

    private Expr Call()
    {
        var e = Primary();
        Token? tok;
        while ((tok = Match(TokenType.LParen, TokenType.Dot)) != null) {
            switch (tok.Value.Type) {
                case TokenType.LParen:
                    e = CallRest(e, tok.Value);
                    break;
                case TokenType.Dot:
                    Token name = Require("property name", TokenType.Ident);
                    e = new PropertyGet(e, new Var(name.Lexeme, name.Location),
                      tok.Value.Location);
                    break;
                default:
                    throw new InvalidOperationException(
                      "internal error: impossible result from Match");
            }
        }
        return e;
    }

    private Expr CallRest(Expr callee, Token lparen)
    {
        List<Expr> args = new List<Expr>();
        if (!Check(TokenType.RParen))
            do
                args.Add(Expression());
            while (Match(TokenType.Comma) != null);

        var rparen = Require("expected ')'", TokenType.RParen);

        /* TODO: revisit this whole idea - do we need it?
         * isn't the VM a stack machine?
         */
        if (args.Count > 255)
            _onError.Error(rparen.Location, $" at {rparen.Lexeme}",
              "more than 255 arguments");

        /* which Location?
         *   book sez rparen (low energy)
         *   callee.Location is sensible choice
         *   I choose lparen (that's where the call "begins" IMO
         */
        return new Call(callee, args, lparen.Location);
    }

    private Expr Primary()
    {
        Token? tok;

        if ((tok = Match(TokenType.Nil)) != null)
            return new Literal(null, tok.Value.Location);

        if ((tok = Match(TokenType.True)) != null)
            return new Literal(true, tok.Value.Location);

        if ((tok = Match(TokenType.False)) != null)
            return new Literal(false, tok.Value.Location);

        if ((tok = Match(TokenType.StrLit, TokenType.NumLit)) != null)
            return new Literal(tok.Value.Literal, tok.Value.Location);

        if ((tok = Match(TokenType.This)) != null)
            return new This(tok.Value.Location);

        if ((tok = Match(TokenType.Ident)) != null)
            return new Var(tok.Value.Lexeme, tok.Value.Location);

        if ((tok = Match(TokenType.Super)) != null) {
            Require("expected '.'", TokenType.Dot);
            Token methodName = Require("expected superclass method name",
              TokenType.Ident);
            return new SuperGet(
              new Var(methodName.Lexeme, methodName.Location),
              tok.Value.Location);
        }

        if (Match(TokenType.LParen) != null) {
            var e = Expression();
            Require("expected ')'", TokenType.RParen);
            return e;
        }

        Require("expected an expression");
        // unreachable
        throw new InvalidOperationException(
          "internal error - should be unreachable");
    }

    private Token? Match(params TokenType[] types)
    {
        foreach (var type in types) {
            if (Check(type)) {
                var tok = _tokens.Next;
                _tokens.Advance();
                return tok;
            }
        }
        return null;
    }

    private Token Require(string message, params TokenType[] types)
    {
        if (_tokens.Next == null) {
            _onError.Error(null, " at end of input", message);
            throw new ParseError();
        }

        Token tok = _tokens.Next!.Value;
        _tokens.Advance();
        foreach (var type in types) {
            if (tok.Type == type)
                return tok;
        }

        _onError.Error(tok.Location, $" at '{tok.Lexeme}'", message);
        throw new ParseError();
    }

    private bool AtEOF => _tokens.Next == null;

    private void RequireEOF()
    {
        if (!AtEOF) {
            _onError.Error(_tokens.Next!.Value.Location,
              $" at \"{_tokens.Next!.Value.Lexeme}\"",
              "expected end of input");
            throw new ParseError();
        }
    }

    private bool Check(TokenType type) => _tokens.Next?.Type == type;

    // seek to "sync point" (see ch6)
    private void Recover()
    {
        while (true) {
            if (Match(TokenType.Semicolon) != null)
                return;

            _tokens.Advance();

            if (_tokens.Next == null)
                return;

            switch (_tokens.Next.Value.Type) {
                case TokenType.Class:
                case TokenType.Fun:
                case TokenType.Var:
                case TokenType.For:
                case TokenType.If:
                case TokenType.While:
                case TokenType.Print:
                case TokenType.Return:
                    return;
                default:
                    break;
            }
        }
    }

    private class ParseError: Exception { }

    private static UnOp TokenUnOp(Token t) => t.Type switch {
        TokenType.Minus => UnOp.Negate,
        TokenType.Not => UnOp.Not,
        _ => throw new ArgumentException($"Invalid unary operator {t}"),
    };

    private static BinOp TokenBinOp(Token t) => t.Type switch {
        TokenType.Plus => BinOp.Plus,
        TokenType.Minus => BinOp.Minus,
        TokenType.Star => BinOp.Times,
        TokenType.Slash => BinOp.DividedBy,
        TokenType.NotEq => BinOp.NotEq,
        TokenType.EqEq => BinOp.EqEq,
        TokenType.Gt => BinOp.Gt,
        TokenType.GtEq => BinOp.GtEq,
        TokenType.Lt => BinOp.Lt,
        TokenType.LtEq => BinOp.LtEq,
        TokenType.Or => BinOp.Or,
        TokenType.And => BinOp.And,
        _ => throw new ArgumentException($"Invalid binary operator {t}"),
    };

    private PeekableStream _tokens;
    private ErrorReporter _onError;
}

File: loxide/src/parser.rs
use std::iter::Peekable;

use crate::{
    error::{self, Error, ErrorBundle, ErrorDetails,},
    lexer::{Token, TokenKind,},
    syntax::*,
    value::Value,
};

pub struct Parser<'a, I: Iterator<Item=error::Result<Token<'a>>>> {
    tokens: Peekable<I>,
    errors: ErrorBundle,
}

macro_rules! match_token {
    ($self:ident, $($p:pat),*) => {
        if let Some(Ok(t)) = $self.tokens.peek() {
            match t.kind {
                $($p => {
                    // safe because of if let ^^^
                    Some($self.tokens.next().unwrap().unwrap())
                },)*
                _ => None,
            }
        } else {
            None
        }
    };
}

macro_rules! check_token {
    ($self:ident, $($p:pat),*) => {
        if let Some(Ok(t)) = $self.tokens.peek() {
            match t.kind {
                $($p => true,)*
                _ => false,
            }
        } else {
            false
        }
    };
}

macro_rules! fold_bin_op {
    ($name:ident, $parser:ident $(,$p:pat)*) => {
        fn $name(&mut self) -> Option<Expr<String>> {
            let mut lhs = self.$parser()?;
            while let Some(tok) = match_token!(self, $($p),*) {
                let rhs = self.$parser()?;
                lhs = Expr::BinOpApp(token_bin_op(&tok).unwrap(),
                  Box::new(lhs), Box::new(rhs), tok.loc);
            }
            Some(lhs)
        }
    };
}

macro_rules! require {
    ($self:ident, $msg:literal $(,$p:pat)*) => {
        match $self.tokens.next() {
            $(Some(Ok(tok @ Token { kind: $p, .. })) => {
                tok
            },)*

            Some(Ok(tok)) => {
                $self.errors.push(Error {
                    loc: Some(tok.loc),
                    lexeme: Some(tok.lexeme.to_string()),
                    details: ErrorDetails::ParseExpected($msg),
                });
                return None;
            },

            Some(Err(e)) => {
                $self.errors.push(e);
                return None;
            },

            None => {
                $self.errors.push(Error {
                    loc: None,
                    lexeme: None,
                    details: ErrorDetails::ParseExpected($msg),
                });
                return None;
            },
        }
    };
}

// TODO: refactor for less duplication with require
macro_rules! require_extract {
    ($self:ident, $msg:literal, $p:pat, $e:expr) => {
        match $self.tokens.next() {
            Some(Ok(Token { kind: $p, .. })) => {
                $e
            },

            Some(Ok(tok)) => {
                $self.errors.push(Error {
                    loc: Some(tok.loc),
                    lexeme: Some(tok.lexeme.to_string()),
                    details: ErrorDetails::ParseExpected($msg),
                });
                return None;
            },

            Some(Err(e)) => {
                $self.errors.push(e);
                return None;
            },

            None => {
                $self.errors.push(Error {
                    loc: None,
                    lexeme: None,
                    details: ErrorDetails::ParseExpected($msg),
                });
                return None;
            },
        }
    };
}

impl<'a, I: Iterator<Item=error::Result<Token<'a>>>> Parser<'a, I> {
    pub fn new(tokens: I) -> Self {
        Self {
            tokens: tokens.peekable(),
            errors: ErrorBundle::new(),
        }
    }

    #[inline]
    fn declaration(&mut self) -> Option<Stmt<String>> {
        if let Some(tok) = match_token!(self, TokenKind::Var) {
            return self.var_decl_rest(tok);
        }

        if let Some(tok) = match_token!(self, TokenKind::Fun) {
            return self.fundef_or_method_rest(tok);
        }

        self.statement()
    }

    fn statement(&mut self) -> Option<Stmt<String>> {
        if let Some(tok) = match_token!(self, TokenKind::If) {
            return self.if_else_rest(tok);
        }

        if let Some(tok) = match_token!(self, TokenKind::While) {
            return self.while_rest(tok);
        }

        if let Some(tok) = match_token!(self, TokenKind::For) {
            return self.for_rest(tok);
        }

        if let Some(tok) = match_token!(self, TokenKind::Print) {
            let e = self.expression()?;
            require!(self, "';'", TokenKind::Semicolon);
            return Some(Stmt::Print(e, tok.loc));
        }

        if let Some(tok) = match_token!(self, TokenKind::Return) {
            if let Some(_) = match_token!(self, TokenKind::Semicolon) {
                return Some(Stmt::Return(None, tok.loc));
            }
            let e = self.expression()?;
            require!(self, "';'", TokenKind::Semicolon);
            return Some(Stmt::Return(Some(e), tok.loc));
        }

        if let Some(tok) = match_token!(self, TokenKind::LBrace) {
            let body = self.block_rest()?;
            return Some(Stmt::Block(body, tok.loc));
        }

        self.expression_statement()
    }

    fn var_decl_rest(&mut self, var: Token) -> Option<Stmt<String>> {
        let name = require_extract!(self,
          "identifier", TokenKind::Ident(name), name.to_string());
        let init = if let Some(_) = match_token!(self, TokenKind::Eq) {
            Some(self.expression()?)
        } else {
            None
        };
        require!(self, "';'", TokenKind::Semicolon);
        Some(Stmt::VarDecl(name, init, var.loc))
    }

    fn fundef_or_method_rest(&mut self, fun: Token) -> Option<Stmt<String>> {
        let name = require_extract!(self,
          "identifier", TokenKind::Ident(name), name.to_string());
        require!(self, "'('", TokenKind::LParen);

        let mut params = Vec::new();
        if !check_token!(self, TokenKind::RParen) {
            loop {
                params.push(require_extract!(self, "parameter name",
                  TokenKind::Ident(name), name.to_string()));
                if match_token!(self, TokenKind::Comma).is_none() {
                    break;
                }
            }
        }

        let rparen = require!(self, "')'", TokenKind::RParen);
        if params.len() > 255 {
            self.errors.push(Error {
                loc: Some(rparen.loc),
                lexeme: Some(rparen.lexeme.to_string()),
                details: ErrorDetails::TooManyArgs,
            });
            return None;
        }

        require!(self, "'{'", TokenKind::LBrace);
        let body = self.block_rest()?;

        Some(Stmt::FunDef(name, params, body, fun.loc))
    }

    fn if_else_rest(&mut self, t_if: Token) -> Option<Stmt<String>> {
        require!(self, "'('", TokenKind::LParen);
        let cond = self.expression()?;
        require!(self, "')'", TokenKind::LParen);
        let s_if = Box::new(self.statement()?);
        let s_else = if let Some(_) = match_token!(self, TokenKind::Else) {
            Some(Box::new(self.statement()?))
        } else {
            None
        };
        Some(Stmt::IfElse(cond, s_if, s_else, t_if.loc))
    }

    fn while_rest(&mut self, t_while: Token) -> Option<Stmt<String>> {
        require!(self, "'('", TokenKind::LParen);
        let cond = self.expression()?;
        require!(self, "')'", TokenKind::LParen);
        let body = Box::new(self.statement()?);
        Some(Stmt::While(cond, body, t_while.loc))
    }

    fn for_rest(&mut self, t_for: Token) -> Option<Stmt<String>> {
        require!(self, "'('", TokenKind::LParen);

        let init = if let Some(_) = match_token!(self, TokenKind::Semicolon) {
            None
        } else if let Some(tok) = match_token!(self, TokenKind::Var) {
            Some(self.var_decl_rest(tok)?)
        } else {
            Some(self.expression_statement()?)
        };

        let cond = if let Some(tok) = match_token!(self, TokenKind::Semicolon) {
            Expr::Literal(Value::Bool(true), tok.loc)
        } else {
            let e = self.expression()?;
            require!(self, "';'", TokenKind::Semicolon);
            e
        };

        let incr = if let Some(_) = match_token!(self, TokenKind::RParen) {
            None
        } else {
            let e = self.expression()?;
            require!(self, "')'", TokenKind::RParen);
            Some(e)
        };

        let body = self.statement()?;
        let body = if let Some(incr) = incr {
            let b_loc = *body.location();
            let i_loc = *incr.location();
            Stmt::Block(vec![body, Stmt::Expr(incr, i_loc)], b_loc)
        } else {
            body
        };
        let body = Stmt::While(cond, Box::new(body), t_for.loc);
        let body = match init {
            Some(init) => {
                let loc = *init.location();
                Stmt::Block(vec![init, body], loc)
            },
            None => body,
        };
        Some(body)
    }

    fn block_rest(&mut self) -> Option<Vec<Stmt<String>>> {
        let mut body = Vec::new();
        while !check_token!(self, TokenKind::RBrace) && !self.is_eof() {
            body.push(self.declaration()?);
        }
        require!(self, "'}'", TokenKind::RBrace);
        Some(body) // once told me
    }

    #[inline]
    fn expression_statement(&mut self) -> Option<Stmt<String>> {
        let e = self.expression()?;
        require!(self, "';'", TokenKind::Semicolon);
        let loc = *e.location();
        Some(Stmt::Expr(e, loc))
    }

    #[inline]
    fn expression(&mut self) -> Option<Expr<String>> {
        self.assignment()
    }

    fn assignment(&mut self) -> Option<Expr<String>> {
        let e = self.logic_or()?;
        if let Some(tok) = match_token!(self, TokenKind::Eq) {
            let val = self.assignment()?;
            match e {
                Expr::Var(v, loc) =>
                    Some(Expr::Assign(v, Box::new(val), loc)),
                _ => {
                    self.errors.push(Error {
                        loc: Some(tok.loc),
                        lexeme: Some(tok.lexeme.to_string()),
                        details: ErrorDetails::NotLValue(format!("{}", val)),
                    });
                    return None;
                }
            }
        } else {
            Some(e)
        }
    }

    fold_bin_op!(logic_or, logic_and, TokenKind::Or);
    fold_bin_op!(logic_and, equality, TokenKind::And);
    fold_bin_op!(equality, comparison, TokenKind::EqEq | TokenKind::NotEq);
    fold_bin_op!(comparison, term,
      TokenKind::Lt | TokenKind::LtEq | TokenKind::Gt | TokenKind::GtEq);
    fold_bin_op!(term, factor, TokenKind::Plus | TokenKind::Minus);
    fold_bin_op!(factor, unary, TokenKind::Star | TokenKind::Slash);

    fn unary(&mut self) -> Option<Expr<String>> {
        if let Some(tok) =
          match_token!(self, TokenKind::Not | TokenKind::Minus) {
            let rhs = self.unary()?;
            Some(Expr::UnOpApp(
              token_un_op(&tok).unwrap(), Box::new(rhs), tok.loc))
        } else {
            self.call()
        }
    }

    #[inline]
    fn call(&mut self) -> Option<Expr<String>> {
        let mut e = self.primary()?;
        while let Some(lparen) = match_token!(self, TokenKind::LParen) {
            e = self.call_rest(e, lparen)?;
        }
        Some(e)
    }

    fn call_rest(&mut self, callee: Expr<String>, lparen: Token
      ) -> Option<Expr<String>> {
        let mut args = Vec::new();
        if !check_token!(self, TokenKind::RParen) {
            loop {
                args.push(self.expression()?);
                if match_token!(self, TokenKind::Comma).is_none() {
                    break;
                }
            }
        }

        let rparen = require!(self, "')'", TokenKind::RParen);
        if args.len() > 255 {
            self.errors.push(Error {
                loc: Some(rparen.loc),
                lexeme: Some(rparen.lexeme.to_string()),
                details: ErrorDetails::TooManyArgs,
            });
            return None;
        }

        Some(Expr::Call(Box::new(callee), args, lparen.loc))
    }

    fn primary(&mut self) -> Option<Expr<String>> {
        if let Some(tok) = match_token!(self,
          TokenKind::Nil | TokenKind::True | TokenKind::False |
          TokenKind::StrLit(_) | TokenKind::NumLit(_)) {
            return Some(Expr::Literal(token_literal(&tok).unwrap(), tok.loc));
        }

        if let Some(Token { kind: TokenKind::Ident(name), loc, .. }) =
          match_token!(self, TokenKind::Ident(_)) {
            return Some(Expr::Var(name.to_string(), loc));
        }

        if let Some(_) = match_token!(self, TokenKind::LParen) {
            let e = self.expression();
            require!(self, "')'", TokenKind::RParen);
            return e;
        }

        require!(self, "an expression")
    }

    #[inline]
    fn is_eof(&mut self) -> bool {
        self.tokens.peek().is_none()
    }

    fn eof(&mut self) -> Option<()> {
        match self.tokens.next() {
            None => Some(()),

            Some(Err(e)) => {
                self.errors.push(e);
                None
            },

            Some(Ok(tok)) => {
                self.errors.push(Error {
                    loc: Some(tok.loc),
                    lexeme: Some(tok.lexeme.to_string()),
                    details: ErrorDetails::ParseExpected("end of input"),
                });
                None
            },
        }
    }

    // recover to a ; or the point before a keyword
    fn recover(&mut self) {
        loop {
            match self.tokens.next() {
                Some(Ok(Token { kind: TokenKind::Semicolon, .. })) => return,
                Some(Err(e)) => { self.errors.push(e) },
                _ => { },
            };

            match self.tokens.peek() {
                None => return,
                Some(Ok(Token {
                    kind: TokenKind::Class
                        | TokenKind::Fun
                        | TokenKind::Var
                        | TokenKind::For
                        | TokenKind::If
                        | TokenKind::While
                        | TokenKind::Print
                        | TokenKind::Return,
                    ..
                })) => return,
                _ => { },
            };
        }
    }
}

fn token_un_op(token: &Token<'_>) -> Option<UnOp> {
    match token.kind {
        TokenKind::Not => Some(UnOp::Complement),
        TokenKind::Minus => Some(UnOp::Negate),
        _ => None,
    }
}

fn token_bin_op(token: &Token<'_>) -> Option<BinOp> {
    match token.kind {
        TokenKind::Plus => Some(BinOp::Add),
        TokenKind::Minus => Some(BinOp::Sub),
        TokenKind::Star => Some(BinOp::Mul),
        TokenKind::Slash => Some(BinOp::Div),
        TokenKind::EqEq => Some(BinOp::Eq),
        TokenKind::NotEq => Some(BinOp::NotEq),
        TokenKind::Lt => Some(BinOp::Lt),
        TokenKind::LtEq => Some(BinOp::LtEq),
        TokenKind::Gt => Some(BinOp::Gt),
        TokenKind::GtEq => Some(BinOp::GtEq),
        TokenKind::And => Some(BinOp::And),
        TokenKind::Or => Some(BinOp::Or),
        _ => None,
    }
}

fn token_literal(token: &Token<'_>) -> Option<Value> {
    match token.kind {
        TokenKind::Nil => Some(Value::Nil),
        TokenKind::True => Some(Value::Bool(true)),
        TokenKind::False => Some(Value::Bool(false)),
        TokenKind::NumLit(n) => Some(Value::Number(n)),
        TokenKind::StrLit(s) => Some(Value::String(s.to_string())),
        _ => None,
    }
}

#[inline]
pub fn parse_expr<'a, I: Iterator<Item=error::Result<Token<'a>>>>(
  tokens: I) -> Result<Expr<String>, ErrorBundle> {
    let mut p = Parser::new(tokens);
    if let Some(e) = p.expression() {
        if let Some(()) = p.eof() {
            return Ok(e);
        }
    }
    Err(p.errors)
}

#[inline]
pub fn parse<'a, I: Iterator<Item=error::Result<Token<'a>>>>(
  tokens: I) -> Result<Vec<Stmt<String>>, ErrorBundle> {
    let mut prog = Vec::new();
    let mut p = Parser::new(tokens);
    let mut ok = true;
    while !p.is_eof() {
        if let Some(s) = p.declaration() {
            prog.push(s);
        } else {
            ok = false;
            p.recover();
        }
    }
    if ok {
        Ok(prog)
    } else {
        Err(p.errors)
    }
}
